Mathlib/Combinatorics/Quiver/Path/Vertices.lean

```diff
@@ -0,0 +1,198 @@
+  simp only [Set.union_singleton]
+  ext v
+  simp only [mem_verticesSet_iff, Set.mem_insert_iff]
+  rw [or_comm]; simp
+
+/-- The length of vertices list equals path length plus one -/
+@[simp]
+lemma vertices_length {V : Type*} [Quiver V] {a b : V} (p : Path a b) :
+    p.vertices.length = p.length + 1 := by
+  induction p with
+  | nil => simp only [vertices_nil, List.length_cons, List.length_nil, zero_add, length_nil]
+  | cons p' e ih =>
+    simp only [vertices_cons, length_cons, List.length_concat, ih]
+
+lemma length_vertices_pos {a b : V} (p : Path a b) :
+    0 < p.vertices.length := by simp
+
+lemma vertices_ne_nil {a : V} {b : V} (p : Path a b) : p.vertices ≠ [] := by
+  simp [← List.length_pos_iff_ne_nil]
+
+/-- The head of the vertices list is the start vertex -/
+@[simp]
+lemma vertices_head? {a b : V} (p : Path a b) : p.vertices.head? = some a := by
+  induction p with
+  | nil => simp only [vertices_nil, List.head?_cons]
+  | cons p' e ih => simp [ih]
+
+@[simp]
+lemma length_toPath {a b : V} (e : a ⟶ b) : e.toPath.length = 1 := rfl
+
+@[simp]
+lemma vertices_getLast {a b : V} (p : Path a b) (h : p.vertices ≠ [] := p.vertices_ne_nil) :
+    p.vertices.getLast h = b := by
+  induction p with
+  | nil => simp only [vertices_nil, List.getLast_singleton]
+  | cons p' e ih => simp
+
+@[simp]
+lemma vertices_comp {a b c : V} (p : Path a b) (q : Path b c) :
+  (p.comp q).vertices = p.vertices.dropLast ++ q.vertices := by
+  induction q with
+  | nil => simpa using (List.dropLast_append_getLast p.vertices_ne_nil).symm
+  | cons q' e ih => simp [ih]
+
+lemma start_mem_vertices {a b : V} (p : Path a b) : a ∈ p.vertices := by
+  induction p with
+  | nil => simp
+  | cons p' e ih => simp [ih]
+
+@[simp] lemma length_eq_zero_iff {a : V} (p : Path a a) :
+    p.length = 0 ↔ p = Path.nil := by
+  cases p <;> tauto
+
+/-- The head of the vertices list is the start vertex. -/
+@[simp]
+lemma vertices_head_eq {a b : V} (p : Path a b) (h : p.vertices ≠ [] := p.vertices_ne_nil) :
+    p.vertices.head h = a := by
+  induction p with
+  | nil => simp only [vertices_nil, List.head_cons]
+  | cons p' _ ih => simp [List.head_append_of_ne_nil (vertices_ne_nil p'), ih]
+
+/-- The last element of the vertices list is the end vertex. -/
+lemma vertices_getLast_eq_end {a b : V} (p : Path a b) :
+  p.vertices.getLast (vertices_ne_nil p) = b := by
+  exact vertices_getLast p (vertices_ne_nil p)
+
+lemma end_prefix_eq_get_vertices {a b c : V} (p₁ : Path a c) (p₂ : Path c b) :
+    c = (p₁.comp p₂).vertices.get
+        ⟨p₁.length, by
+  simp only [vertices_comp, List.length_append, List.length_dropLast,
+    vertices_length, add_tsub_cancel_right, lt_add_iff_pos_right, add_pos_iff,
+    Nat.lt_one_iff, pos_of_gt, or_true]⟩ := by
+  simp only [List.get_eq_getElem, vertices_comp, List.length_dropLast, vertices_length,
+    add_tsub_cancel_right, le_refl, List.getElem_append_right, tsub_self, List.getElem_zero,
+    vertices_head_eq]
```

Comment:
How about we flip the sides? We can also golf the proofs a bit and make this provable by `simp` if we first introduce the following lemma just below `vertices_head?` higher up:
```lean
@[simp]
lemma getElem_vertices_zero {a b : V} (p : Path a b) : p.vertices[0] = a := by
  induction p with
  | nil => simp
  | cons p' e ih => simp [ih]
```
so that this lemma can become
```suggestion
lemma vertices_comp_get_length_eq {a b c : V} (p₁ : Path a c) (p₂ : Path c b)
    (h : p₁.length < (p₁.comp p₂).vertices.length := by simp) :
    (p₁.comp p₂).vertices.get ⟨p₁.length, h⟩ = c := by
  simp
```